src/Core/Services/MultipleMutationInputValidator.cs

```diff
@@ -0,0 +1,575 @@
+            //
+            // derivableColumnsFromRequestBody is initialized with the set of columns whose value is specified in the input (a).
+            Dictionary<string, string> derivableColumnsFromRequestBody = new();
+            foreach ((string backingColumnName, _) in backingColumnData)
+            {
+                derivableColumnsFromRequestBody.TryAdd(backingColumnName, entityName);
+            }
+
+            // When the parent entity is a referenced entity in a relationship, the values of the referencing columns
+            // in the current entity are derived from the insertion in the parent entity. Hence, the input data for
+            // current entity (referencing entity) i.e. derivableColumnsFromRequestBody must not contain values for referencing columns.
+            ValidateAbsenceOfReferencingColumnsInTargetEntity(
+                 columnsInTargetEntity: derivableColumnsFromRequestBody,
+                 derivedColumnsFromSourceEntity: columnsDerivedFromSourceEntity,
+                 nestingLevel: nestingLevel,
+                 targetEntityName: entityName,
+                 metadataProvider: metadataProvider);
+
+            // Add all the columns whose value(s) will be derived from insertion in parent entity to the set of derivable columns (b).
+            foreach (string columnDerivedFromSourceEntity in columnsDerivedFromSourceEntity)
+            {
+                derivableColumnsFromRequestBody.TryAdd(columnDerivedFromSourceEntity, parentEntityName);
+            }
+
+            // For the relationships with the parent entity, where the current entity is a referenced entity,
+            // we need to make sure that we have non-null values for all the referenced columns - since the values for all those
+            // columns will be used for insertion in the parent entity. We can get the referenced column value:
+            // Case 1. Via a scalar value provided in the input for the current entity
+            // Case 2. Via another relationship where a column referenced by the parent of the current entity is a
+            // referencing column for a target entity related with current entity.
+
+            // Case 1: Remove from columnsToBeDerivedFromEntity, the columns which are autogenerated or
+            // have been provided a non-null value in the input for the current entity.
+
+            // As we iterate through columns which can derived from the current entity, we keep removing them from the
+            // columnsToDerive because we know the column value can be derived.
+            // At the end, if there are still columns yet to be derived, i.e.columnsToDerive.count > 0,
+            // we throw an exception.
+            foreach (string columnToBeDerivedFromEntity in columnsToDerive)
+            {
+                if (sourceDefinition.Columns[columnToBeDerivedFromEntity].IsAutoGenerated)
```

Comment:
How about columns with default values?

Reply:
Supporting column with default values does not make sense because:

1. We only know the column has a default value. But we need to get the exact value to be sure that the column value is not null. This requires additional metadata collection.
2. Is not a very useful/intuitive/probable to occur scenario. If you're doing a multiple-create, what makes sense is to provide data/values for fields which are related. (obviously when the field is not autogenerated).
3. The order determination logic gets unnecessarily complicated.